src/ray/gcs/store_client/redis_store_client.cc

```diff
@@ -100,37 +87,273 @@ Status RedisStoreClient::AsyncGet(const std::string &table_name, const std::stri
+  }
+}
+
+Status RedisStoreClient::RedisScanner::ScanKeysAndValues(
+    const ItemCallback<std::unordered_map<std::string, std::string>> &callback) {
+  auto on_done = [this, callback](const Status &status,
+                                  const std::vector<std::string> &result) {
+    if (result.empty()) {
+      callback(std::unordered_map<std::string, std::string>());
+    } else {
+      ScanValues(result, callback);
+    }
+  };
+  return ScanKeys(on_done);
+}
+
+Status RedisStoreClient::RedisScanner::ScanKeys(
+    const MultiItemCallback<std::string> &callback) {
+  auto on_done = [this, callback](const Status &status) {
+    std::vector<std::string> result;
+    result.insert(result.begin(), keys_.begin(), keys_.end());
+    callback(status, result);
+  };
+  Scan(on_done);
+  return Status::OK();
+}
+
+void RedisStoreClient::RedisScanner::Scan(const StatusCallback &callback) {
+  if (shard_to_cursor_.empty()) {
+    callback(Status::OK());
+    return;
+  }
+
+  size_t batch_count = RayConfig::instance().maximum_gcs_scan_batch_size();
+  for (const auto &item : shard_to_cursor_) {
+    ++pending_request_count_;
+
+    size_t shard_index = item.first;
+    size_t cursor = item.second;
+
+    auto scan_callback = [this, shard_index,
+                          callback](const std::shared_ptr<CallbackReply> &reply) {
+      OnScanCallback(shard_index, reply, callback);
+    };
+
+    // Scan by prefix from Redis.
+    std::vector<std::string> args = {"SCAN",  std::to_string(cursor),
+                                     "MATCH", match_pattern_,
+                                     "COUNT", std::to_string(batch_count)};
+    auto shard_context = redis_client_->GetShardContexts()[shard_index];
+    Status status = shard_context->RunArgvAsync(args, scan_callback);
+
+    if (!status.ok()) {
+      RAY_LOG(FATAL) << "Scan failed, status " << status.ToString();
+    }
+  }
+}
+
+void RedisStoreClient::RedisScanner::OnScanCallback(
+    size_t shard_index, const std::shared_ptr<CallbackReply> &reply,
+    const StatusCallback &callback) {
+  RAY_CHECK(reply);
+  std::vector<std::string> scan_result;
+  size_t cursor = reply->ReadAsScanArray(&scan_result);
+
+  // Update shard cursors and keys_.
+  {
+    absl::MutexLock lock(&mutex_);
+    auto shard_it = shard_to_cursor_.find(shard_index);
+    RAY_CHECK(shard_it != shard_to_cursor_.end());
+    if (cursor == 0) {
+      shard_to_cursor_.erase(shard_it);
+    } else {
+      shard_it->second = cursor;
+    }
+
+    keys_.insert(scan_result.begin(), scan_result.end());
+  }
+
+  // Scan result is empty, continue scan.
+  if (--pending_request_count_ == 0) {
+    Scan(callback);
+  }
+}
+
+void RedisStoreClient::RedisScanner::ScanValues(
+    const std::vector<std::string> &keys,
+    const ItemCallback<std::unordered_map<std::string, std::string>> &callback) {
+  for (const auto &key : keys) {
+    ++pending_read_count_;
+
+    std::vector<std::string> args = {"GET", key};
```

Comment:
use `mget` to reduce number of requests.

Reply:
fixed